kbfsgit/runner.go

```diff
@@ -925,40 +988,143 @@ func (r *runner) handleFetchBatch(ctx context.Context, args [][]string) (
 	return err
 }
 
-// handlePushBatch: From https://git-scm.com/docs/git-remote-helpers
-//
-// push +<src>:<dst>
-// Pushes the given local <src> commit or branch to the remote branch
-// described by <dst>. A batch sequence of one or more push commands
-// is terminated with a blank line (if there is only one reference to
-// push, a single push command is followed by a blank line). For
-// example, the following would be two batches of push, the first
-// asking the remote-helper to push the local ref master to the remote
-// ref master and the local HEAD to the remote branch, and the second
-// asking to push ref foo to ref bar (forced update requested by the
-// +).
-//
-// push refs/heads/master:refs/heads/master
-// push HEAD:refs/heads/branch
-// \n
-// push +refs/heads/foo:refs/heads/bar
-// \n
-//
-// Zero or more protocol options may be entered after the last push
-// command, before the batch’s terminating blank line.
-//
-// When the push is complete, outputs one or more ok <dst> or error
-// <dst> <why>? lines to indicate success or failure of each pushed
-// ref. The status report output is terminated by a blank line. The
-// option field <why> may be quoted in a C style string if it contains
-// an LF.
-func (r *runner) handlePushBatch(ctx context.Context, args [][]string) (
-	err error) {
-	repo, fs, err := r.initRepoIfNeeded(ctx, gitCmdPush)
+// canPushAll returns true if a) the KBFS repo is currently empty, and
+// b) we've been asked to push all the local references (i.e.,
+// --all/--mirror).
+func (r *runner) canPushAll(
+	ctx context.Context, repo *gogit.Repository, args [][]string) (
+	canPushAll, repoEmpty bool, err error) {
+	refs, err := repo.References()
 	if err != nil {
+		return false, false, err
+	}
+	defer refs.Close()
+
+	for {
+		ref, err := refs.Next()
+		if errors.Cause(err) == io.EOF {
+			break
+		} else if err != nil {
+			return false, false, err
+		}
+
+		if ref.Type() != plumbing.SymbolicReference {
+			r.log.CDebugf(ctx, "Remote has at least one non-symbolic ref: %s",
+				ref.String())
+			return false, false, nil
+		}
+	}
+
+	sources := make(map[string]bool)
+	for _, push := range args {
+		if len(push) != 1 {
+			return false, false, errors.Errorf("Bad push request: %v", push)
+		}
+		refspec := gogitcfg.RefSpec(push[0])
+		// If some ref is being pushed to a different name on the
+		// remote, we can't do a push-all.
+		if refspec.Src() != refspec.Dst("").String() {
```

Comment:
I wanted to know why `""` was used here, but I'm having a hard time understanding the `Dst` method [here](https://github.com/src-d/go-git/blob/f9879dd043f84936a1f8acb8a53b74332a7ae135/config/refspec.go#L114). Is this the correct test for ref name being same or not?

Reply:
The argument is only used for wildcard specs, and then it's used to figure out if the destination branch you're interested in matches the spec.  We don't have wildcare specs, so we don't need to worry about it.

Reply:
I see; is the argument a branch name, or a remote name?

Reply:
Branch name.